ta/pkcs11/src/token_capabilities.c

```diff
@@ -0,0 +1,240 @@
+	 *				Wr|Uw: Wrap/Unwrap -------.   |  |  |
+	 *				Dr: Derive ----------.    |   |  |  |
+	 *		Sr|Vr: SignRecover/VerifyRecov --.   |    |   |  |  |
+	 *		Si|Ve: Sign/Verify --------.     |   |    |   |  |  |
+	 *		En|De: Encrypt/Decrypt     |     |   |    |   |  |  |
+	 *		Di: Digest -----.    |     |     |   |    |   |  |  |
+	 *				|   / \   / \   / \  |   / \  |  |  |
+	 * Mechanism			Di|En|De|Si|Ve|Sr|Vr|Dr|Wr|Uw|Ge|Gp|1
+	 */
+	MECHA(AES_ECB,			0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0),
+	/*
+	 * Mechanism			 Di|En|De|Si|Ve|Sr|Vr|Dr|Wr|Uw|Ge|Gp|1
+	 *                               |   \_/   \_/   \_/  |   \_/  |  |  |
+	 *		Di: Digest ------'    |     |     |   |    |   |  |  |
+	 *		En|De: Encrypt/Decrypt'     |     |   |    |   |  |  |
+	 *		Si|Ve: Sign/Verify ---------'     |   |    |   |  |  |
+	 *		Sr|Vr: SignUpdate/VerifyRecover --'   |    |   |  |  |
+	 *				Dr: Derive -----------'    |   |  |  |
+	 *				Wr|Uw: Wrap/Unwrap --------'   |  |  |
+	 *				Ge: Generate secret value -----'  |  |
+	 *				Gp: Generate secret pair ---------'  |
+	 *				1: One shot processing only ---------'
+	 */
+};
+
+#if CFG_TEE_TA_LOG_LEVEL > 0
+const char *mechanism_string_id(enum pkcs11_mechanism_id id)
+{
+	const size_t offset = sizeof("PKCS11_CKM_") - 1;
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs11_modes); n++)
+		if (pkcs11_modes[n].id == id)
+			return pkcs11_modes[n].string + offset;
+
+	return "Unknown ID";
+}
+#endif /*CFG_TEE_TA_LOG_LEVEL*/
+
+/*
+ * Return true if @id is a valid mechanism ID
+ */
+bool mechanism_is_valid(enum pkcs11_mechanism_id id)
+{
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs11_modes); n++)
+		if (id == pkcs11_modes[n].id)
+			return true;
+
+	return false;
+}
+
+/*
+ * Return true if mechanism ID is valid and flags matches PKCS#11 compliancy
+ */
+bool mechanism_flags_complies_pkcs11(uint32_t mechanism_type, uint32_t flags)
+{
+	uint32_t test_flags = 0;
+	size_t n = 0;
+
+	test_flags = flags & (PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT |
+			      PKCS11_CKFM_DERIVE | PKCS11_CKFM_DIGEST |
+			      PKCS11_CKFM_SIGN | PKCS11_CKFM_SIGN_RECOVER |
+			      PKCS11_CKFM_VERIFY | PKCS11_CKFM_VERIFY_RECOVER |
+			      PKCS11_CKFM_GENERATE |
+			      PKCS11_CKFM_GENERATE_KEY_PAIR |
+			      PKCS11_CKFM_WRAP | PKCS11_CKFM_UNWRAP);
+
+	for (n = 0; n < ARRAY_SIZE(pkcs11_modes); n++) {
+		if (pkcs11_modes[n].id == mechanism_type) {
+			if (test_flags & ~pkcs11_modes[n].flags)
+				EMSG("%s flags: 0x%"PRIx32" vs 0x%"PRIx32,
+				     id2str_mechanism(mechanism_type),
+				     test_flags, pkcs11_modes[n].flags);
+
+			return (test_flags & ~pkcs11_modes[n].flags) == 0;
+		}
+	}
+
+	/* Mechanism ID unexpectedly not found */
+	return false;
+}
+
+/*
+ * struct supported_mechanism - Describe mechanism supported by the PKCS11 TA
+ * @id: ID of the mechanism
+ * @flags: PKCS11_CKFM_* flags supported tby the TA for the mechanism
+ * @string: Helper string of a mechanism ID for debug purpose
+ */
+struct supported_mechanism {
```

Comment:
This looks much like `struct pkcs11_mechachism_modes`, any chance to merge them instead?

Reply:
Yes, right, will do :)